metrics/memstats.go

```diff
@@ -0,0 +1,539 @@
+	//
+	// There is no StackIdle because unused stack spans are
+	// returned to the heap (and hence counted toward HeapIdle).
+	StackInuse *stats.Int64Measure
+
+	// StackSys is bytes of stack memory obtained from the OS.
+	//
+	// StackSys is StackInuse, plus any memory obtained directly
+	// from the OS for OS thread stacks (which should be minimal).
+	StackSys *stats.Int64Measure
+
+	// MSpanInuse is bytes of allocated mspan structures.
+	MSpanInuse *stats.Int64Measure
+
+	// MSpanSys is bytes of memory obtained from the OS for mspan
+	// structures.
+	MSpanSys *stats.Int64Measure
+
+	// MCacheInuse is bytes of allocated mcache structures.
+	MCacheInuse *stats.Int64Measure
+
+	// MCacheSys is bytes of memory obtained from the OS for
+	// mcache structures.
+	MCacheSys *stats.Int64Measure
+
+	// BuckHashSys is bytes of memory in profiling bucket hash tables.
+	BuckHashSys *stats.Int64Measure
+
+	// GCSys is bytes of memory in garbage collection metadata.
+	GCSys *stats.Int64Measure
+
+	// OtherSys is bytes of memory in miscellaneous off-heap
+	// runtime allocations.
+	OtherSys *stats.Int64Measure
+
+	// NextGC is the target heap size of the next GC cycle.
+	//
+	// The garbage collector's goal is to keep HeapAlloc ≤ NextGC.
+	// At the end of each GC cycle, the target for the next cycle
+	// is computed based on the amount of reachable data and the
+	// value of GOGC.
+	NextGC *stats.Int64Measure
+
+	// LastGC is the time the last garbage collection finished, as
+	// nanoseconds since 1970 (the UNIX epoch).
+	LastGC *stats.Int64Measure
+
+	// PauseTotalNs is the cumulative nanoseconds in GC
+	// stop-the-world pauses since the program started.
+	//
+	// During a stop-the-world pause, all goroutines are paused
+	// and only the garbage collector can run.
+	PauseTotalNs *stats.Int64Measure
+
+	// NumGC is the number of completed GC cycles.
+	NumGC *stats.Int64Measure
+
+	// NumForcedGC is the number of GC cycles that were forced by
+	// the application calling the GC function.
+	NumForcedGC *stats.Int64Measure
+
+	// GCCPUFraction is the fraction of this program's available
+	// CPU time used by the GC since the program started.
+	//
+	// GCCPUFraction is expressed as a number between 0 and 1,
+	// where 0 means GC has consumed none of this program's CPU. A
+	// program's available CPU time is defined as the integral of
+	// GOMAXPROCS since the program started. That is, if
+	// GOMAXPROCS is 2 and a program has been running for 10
+	// seconds, its "available CPU" is 20 seconds. GCCPUFraction
+	// does not include CPU time used for write barrier activity.
+	//
+	// This is the same as the fraction of CPU reported by
+	// GODEBUG=gctrace=1.
+	GCCPUFraction *stats.Float64Measure
+}
+
+// Start initiates a Go routine that starts pushing metrics into
+// the provided measures.
+func (msp *MemStatsProvider) Start(ctx context.Context, period time.Duration) {
+	go func() {
+		ticker := time.NewTicker(period)
```

Comment:
```
for time.NewTicker(period) {
}
```

Reply:
That doesn't work with cancellation?

Reply:
It does, but I guess it does not simply things